### DISCUSSION TITLE
How to omit empty string by default

### DISCUSSION THREAD

**Discussion by xsjcTony** [react-hook-form/react-hook-form#10527] [Q&A]
Consider about this:
```ts
const { control } = useForm<{ description?: string }>({
  defaultValue: {
    // have to do this, since I'm using controlled input, otherwise React will warn. Not a fan of `register`
    description: ''
  }
}
```
But when I'm submitting the form, I will get `{ "description": "" }` even if I didn't input anything.
I know it's because I've set the default value but the reason is just what I described above.

Is there anyway to remove empty string by default?

**Comment by leapful** [comment]
@xsjcTony 

You can use a 3rd party library like zod or yup to transform the value.

Working demo: https://codesandbox.io/s/react-hook-form-zod-resolver-ts-example-forked-mkj52w?file=/src/App.tsx

**Reply by xsjcTony** [reply]
Yeah I'm aware I can do that, but it's kind of a nightmare if I need to apply this transform to every `optional` field. If I apply transform to the whole `object schema` then it's gonna be another nightmare to deal with the type system, especially in `zod`.

That's why I want to know whether there's a way to transform the value after this library parsed the inputs and just before submission.

**Reply by xsjcTony** [reply]
https://github.com/colinhacks/zod/issues/2524 It's just really a nightmare to work around this kind of things...

**Reply by leapful** [reply]
@xsjcTony 

You can just use a extended schema to parse the form values on valid or on error of `handleSubmit` via `getValues` method.

Working demo: https://codesandbox.io/s/react-hook-form-zod-resolver-ts-example-forked-qjfq9m

**Reply by xsjcTony** [reply]
Well yeah, that's working but it's still lying to TypeScript.

- The output type of `extendedSchema.parse()` will be `{}` which is assignable to almost everything
- `Object.entries().reduce()` just simply wiped out all key-value pair constrained type initially defined by the original schema
- If in the `reduce` function I use format `acc[key] = value || undefined`, it will straight away fall into a type error, which is expected. (If you cast the original value `{} as Schema`, which is required to correct the type of the output)

And if I want to omit some fields based on some other fields, it's even more complex
- `zod` has no `when` API so it's a little bit frustrating
- `yup` does not provide any type-safety in `when` API

Hence yeah... Although it's working in the runtime, there's no type-safety at all. 😥
That said, I really want to find out a way to do it when guaranteeing type-safety... otherwise I have to lie to the type system somehow...
